### PR TITLE
Core: Introduce new Option class NamedRange

### PR DESCRIPTION
**Pull request by el-u**
The existing option class `SpecialRange` has some shortcomings:
- https://discord.com/channels/731205301247803413/1162383276778995722
  Rolling with values such as "random", "random-low", etc. can lead to exceptions if the class is misconfigured because the developer did not adjust `special_range_cutoff`. This is possible if `range_start` is set to a value larger than `special_range_cutoff` (which has a default value of 0 if not adjusted). Random values are selected from the interval [`special_range_cutoff`, `range_end`], which can then result in values that are lower than `range_start`, which is not allowed. (To make matters worse, `special_range_cutoff` is not documented anywhere.)
- `SpecialRange` only supports special named values that are inside or below the visible range (but not above).

This PR introduces a new options class, `NamedRange`. The design goal was to be able to handle everything that `SpecialRange` could, while also improving upon the two things mentioned above:
- `NamedRange` does not have a `special_range_cutoff` attribute. Only the usual `range_start` and `range_end` need to be configured. Rolling with values such as "random", "random-low", etc. will select values from the interval [`range_start`, `range_end`].
- `NamedRange` can handle any value as a special named one (be it below, inside, or above the visible range).

In short, the new class should provide more features while also being easier to use for everyone.
Therefore, all worlds are migrated to use the new `NamedRange`. (In most cases, this is a drop-in replacement; small adjustments were necessary only if `range_start` was previously negative or if `special_range_cutoff` had been changed from the default value.),
The old class, `SpecialRange`, will be deprecated (to be removed in a future update) because it is difficult to use and does not offer any benefit compared to the `NamedRange` introduced in this PR.

### CODE REVIEW

File: WebHostLib/options.py [hunk 1]
```diff
@@ -84,7 +84,7 @@ def get_html_doc(option_type: type(Options.Option)) -> str:
                     "max": option.range_end,
                 }
 
-                if issubclass(option, Options.SpecialRange):
+                if issubclass(option, Options.NamedRange):
                     game_options[option_name]["type"] = 'special_range'
```
Comment (agilbert1412) [inline WebHostLib/options.py:88]: Shouldn't the javascript-side name also change, to stay in sync?

Comment (agilbert1412) [reply WebHostLib/options.py:88]: This will include a few other places where you need to change it. Take a look at the places I had to change in #2305

**Review by agilbert1412 — CHANGES_REQUESTED** [review]
I must admit that, looking at the implementation, this looks much better than what I had imagined when reading your description of it. Possibly I was misunderstanding your proposition, or maybe you got the idea for a better version later. Regardless, while I originally disagreed, I now agree with this as a way to go about it.
I now realize that the whole point of the cutoff was to allow named values outside the range, so if the SpecialRange simply... allows that, then there is no longer a need for a distinction between NamedRange and SpecialRange.

So I like it

Few questions:
1: It seems like these changes would work fine even if the class stayed named "SpecialRange", yet you renamed it. I do agree that the new name is better, but is it really worth the breaking change in API, and deprecation process, just for this rename?
2: Have you considered the use case where a dev might want to add special not-named values that are outside the range, for the purpose of Easter Eggs, debugging tools or simply secret features? This is the only use case I can think of that this PR is removing, and it is a bit of a stretch, I just want to make sure you're aware.

**Comment by beauxq** [comment]
I tried this with the website.
It's a little weird that the select box can move the slider to a place that the slider can't be moved to manually.
But it's not bad I guess.

File: WebHostLib/options.py (hunk 1 above)
Comment (agilbert1412) [reply WebHostLib/options.py:88]: Yes, but I am suggesting changing the type here, for consistency, and subsequently also adapting the js files to match

Comment (Berserker66) [reply WebHostLib/options.py:88]: I agree that it should be matched on the JS side for consistency. Searching for the term across the project should find its variants, it's nice.

**Comment by NewSoupVi** [comment]
> 2: Have you considered the use case where a dev might want to add special not-named values that are outside the range, for the purpose of Easter Eggs, debugging tools or simply secret features? This is the only use case I can think of that this PR is removing, and it is a bit of a stretch, I just want to make sure you're aware.

Interesting! If I had known this was possible with SpecialRange, I would've actually used this ages ago, as there are some higher values (8-11) for one of my ranges that are possible but require a hard trick, so I chose to limit the range to values 1-7, but have always wondered if there is a way to make 8-11 yaml-only as a hidden advanced setting.

Well, sad that I learn about this now that it's going away, but oh well. :D

**Review by agilbert1412 — APPROVED** [review]

**Comment by NewSoupVi** [comment]
> > Interesting! If I had known this was possible with SpecialRange, I would've actually used this ages ago, as there are some higher values (8-11) for one of my ranges that are possible but require a hard trick, so I chose to limit the range to values 1-7, but have always wondered if there is a way to make 8-11 yaml-only as a hidden advanced setting.
> 
> Actually, this _wasn't_ specially possible with `SpecialRange`. One of the shortcomings of `SpecialRange` (as seen in the first post) is that it can just deal with additional values below and inside the range, but not _above_ it.
> 
> That aside, if you really wanted some undocumented option values, couldn't you simply use `__init__` customization as I mentioned? It might look something like this:
> 
> ```python
> class MyOption(NamedRange):
>     ...
>     def __init__(self, value: int) -> None:
>         if 8 <= value <= 11:  # undocumented advanced difficulty options
>             self.value = value
>         else:
>             super().__init__(value)
> ```

Oh! That's pretty epic.

**Comment by alwaysintreble** [comment]
code lgtm. i don't think there's a good way to handle that specific use case generically, and the ability for a world to implement it itself has always been possible, and something i'd rather encourage for more involved options.

**Review by alwaysintreble — APPROVED** [review]

**Comment by black-sliver** [comment]
Was about to go through this, but we ended up creating merge conflicts in multiple places. Can you try to resolve them when you have time?

File: WebHostLib/static/assets/player-options.js [hunk 2]
```diff
@@ -194,13 +194,13 @@ const buildOptionsTable = (options, romOpts = false) => {
         element.appendChild(randomButton);
         break;
 
-      case 'special_range':
+      case 'named_range':
         element = document.createElement('div');
-        element.classList.add('special-range-container');
+        element.classList.add('named-range-container');
```
Comment (black-sliver) [inline WebHostLib/static/assets/player-options.js:199]: I think you forgot to update the .css file that defines this and `named-range-wrapper`. The layout is broken for me in webhost. I can send a screenshot if it isn't broken for you for some reason.

**Review by black-sliver — COMMENTED** [review]
Changes to games look fine. HTML is somehow broken (read below). Should we ping all affected games' maintainers?

File: WebHostLib/options.py [hunk 3]
```diff
@@ -84,8 +84,8 @@ def get_html_doc(option_type: type(Options.Option)) -> str:
                     "max": option.range_end,
                 }
 
-                if issubclass(option, Options.SpecialRange):
-                    game_options[option_name]["type"] = 'special_range'
```
[the remark this answers is not in the document]
Comment (alwaysintreble) [reply WebHostLib/options.py:88]: since the "new" SpecialRange is a subclass of NamedRange it should still work here. i do agree with hard deprecating it since it's already fixed on source,

**Review by agilbert1412 — APPROVED** [review]

**Comment by BadMagic100** [comment]
The change on the HK side looks correct, thanks for the ping

**Review by axe-y — APPROVED** [review]
For DLC Quest all good

File: WebHostLib/options.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (black-sliver) [reply WebHostLib/options.py:88]: I believe using SpecialRange from an APWorld will work on both 0.4.3 (as before) and 0.4.4 (emulated) as long as it's frozen or run with -O, so APWorlds should only port to NamedRange once we have a 0.4.4 release or pre-release.

**Comment by LegendaryLinux** [comment]
`weighted-options.js` was fetching `weighted-settings.json`, which does not include the new options. We should probably get rid of the `weighted-settings.json` generated file and the `weighted-settings` generated directory, as they are no longer used to my knowledge.

The Weighted Options page does not currently allow values outside the specified range to be chosen. This is a limitation of the `<input type="range"/>` element. I don't know of a good way to represent this information or make it available for selection on the Weighted Options page, so the solution might be to simply continue displaying the un-selectable options as the page currently does, and adding a note that they should be edited in the YAML file directly.

It may also be possible to use an `<input type="number"/>` instead of the span at the end, but that would involve checking for valid inputs, and I'm not sure that effort is worth it, and I suspect including that input may confuse people.

So the options for the weighted-options page seem to be:
1. Add a note explaining that options outside of the available range should be edited directly in the YAML
2. Add an `<input />` to replace the number and let people enter their own values, with input validation
3. Some other option I didn't think of

As for the Player Options page, I don't know of a way to make it cleaner than it is. The select allows the option to be chosen, but the range slider can't reach values outside she specified range. Again, a limitation of the HTML input.

**Comment by LegendaryLinux** [comment]
I don't actually think the player options page needs to be updated for this. The user chooses an option they want, and the value is correctly set. If they move the range slider, it correctly updates to a value within the range. I'd consider adding a note to mention certain options may exist outside the specified range, but that might actually confuse people. The current implementation might actually be the least confusing.

As for the weighted options page, I think it's sufficient to check the number entered is either within the range or is one of the special values. As this just creates a row with a slider not dependent on the specified range, that should be enough. Maybe just add a note about special values outside the range.

**Comment by beauxq** [comment]
I agree with LegendaryLinux. I think the current implementation is better than disabling the slider for out-of-range values.
Disabling the slider would make it a little more work to move away from the out-of-range values.

**Comment by LegendaryLinux** [comment]
If you don't mind me pushing to your branch, I can handle the weighted-settings update this week.

**Comment by ThePhar** [comment]
Adding `release/blocker` as some merged commits can cause generation failures if used on main until this is merged.

**Review by LegendaryLinux — APPROVED** [review]
I have updated weighted-options to work with this change, and added a few UI updates as well.

**Review by ThePhar — APPROVED** [review]
